### ISSUE TITLE
Reporting heterogeneous Biolink Model versions across x-maturity environments.

### ISSUE THREAD

**Issue by RichardBruskiewich** [NCATSTranslator/translator_extensions#21]
As Translator evolves, the KP and ARA server `x-maturity` environments appear to be updated asynchronously with respect to Biolink Model versions. 

For example, as of January 2023, the production environments may still,  represent stable releases (from the fall 2022) of  Biolink Model  release 2.4.8, whereas, emerging `development` and `testing` (CI) endpoints may already be compliant with  Biolink Model 3.1.1.

How can Translator SmartAPI Registry entries encode for this?

**Comment by brettasmi** [comment]
The [OpenAPI server object](https://swagger.io/specification/#server-object) supports additional variables that could be used to specify the Biolink version of that `x-maturity` environment.

**Comment by RichardBruskiewich** [comment]
Here, I will briefly brainstorm on this issues and options that come to mind for its resolution.

1. We could simply start by asking whether or not multiple Biolink Model releases should be allowed (by Translator architecture policy) within a given Registry  or rather, should every `x-maturity` environment Registry entry be expected to be homogeneous with respect to Biolink Model releases?  It is already community policy (right @edeutsch, @cbizon?) that Registry entries (hence, all of their deployed `x-maturity` environments?) are already expected to be globally unique with respect to indexing by the 2-tuple of `infores` and `info.x-trapi.version`.   One solution to Biolink Model release heterogeneity across `x-maturity` environments may be to extend the policy of global uniqueness to index by 3-tuple including the `info.x-translator.biolink-version` property value.
    - **Pro:** doesn't change the Registry schema design, only constrains the way it is applied, in that a new Biolink Model release of a given (KP or ARA) resource would necessitate creation of a new Registry entry, with the same `infores` and `info.x-trapi.version` as the previous entry but with the new `info.x-translator.biolink-version` property value.
    - **Con:** Biolink Model releases evolve somewhat faster than TRAPI resource updates and most of the other Registry metadata is likely unchanged, so creating a new entry for each Biolink version release seems tedious and redundant, perhaps resulting in a confusing proliferation of the entries.
2. The `info.x-translator.biolink-version` property value is currently a static SmartAPI OpenAPI 'extension' value.  Maybe it should not be statically reported? Perhaps reporting it dynamically through the `/meta_knowledge_graph` TRAPI endpoint is a better semantic home for this metadata?
    - **Pro:** it is arguable that the Biolink Model version is actually not really a property of 'Translator' (for publication under `info.x-translator`)  but rather, an intrinsic metadata of the knowledge itself (about which we report various other Biolink model characteristics in the `/meta_knowledge_graph`).  The dynamic nature of this option will better support rapid evolution of versioning of underlying data sources.
    - **Con:** the TRAPI specification would need quick revision (and community implementation) to accommodate this proposed change (albeit, one time pain?). The Biolink Model version would conceivably be less visible at the Registry level (if this is a concern)
3. The `info.x-translator.biolink-version` property value could be retained as a global default for the Registry entry but an mechanism to override it elsewhere in the Registry entry could be specified and implemented. There are a couple of possible locations for this:
    3 a)  the current (recently updated) design of the `info.x-trapi.test_data_location`  property value includes for `x-maturity` JSON object entries, so perhaps, a way forward on this issue is to allow for overriding of the top-level Biolink release here (alongside the 'url' tag of the test_data_location URL entry?)

**Comment by cbizon** [comment]
Yeah I agree with @edeutsch .   @RichardBruskiewich what actual problem is it that we're trying to solve?

**Comment by colleenXu** [comment]
Note that staging = ITRB CI (testing is ITRB test)

**Comment by RichardBruskiewich** [comment]
@edeutsch @cbizon I have to laugh a little bit here, but the problem is exactly what @edeutsch says, that _"...it seems infeasible to me to have everyone on the same Biolink release..."_.  

I just realized that my wording of option 1 above was a bit confusing, though: I am not advocating that all KP and ARA's only implement one Biolink Model version at all cost. Rather, I'm just asking how to consistently **_report_** what release they **_are_** implementing, across all the endpoints they expose (I tried to clarify option 1 text accordingly).

That is, I'm actually stating that we face a more stringent challenge, that _within_ a given KP or ARA, it seems to be infeasible to expect all `x-maturity` environments to be using the same Biolink Model release at a given time.  The current reality is that the resources don't always implement the same Biolink Version across all their `x-maturity` environment (endpoints), let alone, the **info.x-translator.biolink-version** indicated value.  Otherwise put, a given KP may have `production` implemented using the **info.x-translator.biolink-version** (e.g. 2.4.8) but may now publishing Biolink Model release 3.1.1 in 'development' or 'testing' `x-maturity` environments.

In terms of SRI Testing of KP and ARA components for compliance to the Biolink Model, it is important to know which Biolink Model release is being tested, to giving non-spurious validation results; otherwise, the testing with test data that assumes other releases of the Biolink model may report errors which are not relevant.

Maybe this issue is based on a false assumption for testing: that SRI Testing needs to know the Biolink Model release of the target KP or ARA.  At the moment, if the Biolink Model version is not stated as a test run parameter, then code assumes that the Translator SmartAPI Registry **info.x-translator.biolink-version** property value is the testing default. 

Perhaps the target Biolink Model release for testing ought to just be a mandatory (CLI) test run specified parameter, such that the actual Biolink Model version implemented by the target KP or ARA, or reported by the **info.x-translator.biolink-version**  is simply ignored;  rather, we simply decide that "_we want to test 3.1.1. compliance_" and say so in the CLI. If we don't say it, then SRI Testing CLI complains and stops!

So, @edeutsch @cbizon, this is not about tracking sheet project-level compliance but mainly about accurately publicizing the actual Biolink Model release of every endpoint.   For this reason, the various alternatives for refining the reporting of Biolink Model version of endpoints are brainstormed above.

I'm currently suspecting that option 2 - publicizing the implemented Biolink Model release as another reported parameter of the TRAPI` /meta_knowledge_graph` API endpoint - might be the most logical location for this metadata given its nature: it is a (meta)property of the underlying knowledge graph, n'est ce pas?  

However, @edeutsch implementing Option 2 would require a quick revision of TRAPI and earnest collaboration for implementation ASAP (i.e. in TRAPI 1.4).  But, would this not be a super easy fix to TRAPI, easily curated by everyone?

**Comment by edeutsch** [comment]
maybe. I thinking that it would be much easier for everyone to put it in their servers block. This area is not even really strictly part of TRAPI. It is managed by the SmartAPI folks. So we can convince them to do this:
```
servers:
- description: ARAX TRAPI 1.3 endpoint - production
  url: https://arax.transltr.io/api/arax/v1.3
  x-maturity: production
  x-biolink-version: 2.2.11
- description: ARAX TRAPI 1.3 endpoint - development
  url: https://arax.ncats.io/test/api/arax/v1.3
  x-maturity: development
  x-biolink-version: 3.0.3
```
then everyone can be prodded to switch to this completely independently of TRAPI 1.4, which will take months.
So if you want it fast, then the above solution is fast (IMO).
Implementing it in an endpoint will be slow and more work.

**Comment by RichardBruskiewich** [comment]
Voting for option 3 above, I see... ;-))

There seems to be a trade-off between visibility of the metadata and its maintainability. Pushing it into /meta_knowledge_graph seems to bring it closer to the semantic metal. Putting in the servers block makes it more visible (akin to the current **info.x-translator.biolink-version**), and one supposes that TRAPI schema validation would ensure that it is properly set with some value... although the discipline of ensuring that this value remains consistent with the internal knowledge graph version, is still a matter of metadata curation discipline.

**Comment by edeutsch** [comment]
Do we want to bring this up at next week's Architecture call? If we do option 2, this would be a TRAPI issue. But if we do option 3, then it is more a SmartAPI issue for Chunlei and his team. Decide at Arch where to send it @cbizon ?

**Comment by newgene** [comment]
Either option 3 (with some modification) or 4 would be fine with me, and my notes are below:

- I still prefer to option 4 like this:

   ```
  info:
    x-translator: 
      biolink-version:
          default: "2.4.8"
          staging: "3.1.0"
          development: "3.1.1"
   ```
   with
   ```
  info:
    x-translator: 
      biolink-version: "2.4.8"
   ```
   still as a back-compatible option


- But if we decide to go for option 3, I suggest:
  - nested under `server.x-translator` for translator-specific extension fields
  - move other "x-maturity" specific values under this level as well, so we don't mix two styles

   ```
   servers:
     - description: ARAX TRAPI 1.3 endpoint - production
       url: https://arax.transltr.io/api/arax/v1.3
       x-maturity: production
       x-translator:
         - biolink-version: 2.2.11
     - description: ARAX TRAPI 1.3 endpoint - development
       url: https://arax.ncats.io/test/api/arax/v1.3
       x-maturity: development
       x-translator:
         - biolink-version: 3.0.3
   ```

  With this option, we will always set `biolink-version` at the "server" level, won't be able to come back to set it at API level when all servers are on the same version.

**Comment by RichardBruskiewich** [comment]
Option 4 would be totally fine with the caveat that **all** endpoints of a given x-maturity environment will have to comply with **one** Biolink Model version.  

Option 3 **might**, however, provide for greater granularity, down to individual server endpoints, if that is desirable, since some KPs or ARAs may have more than one server of a given x-`maturity` environment (e.g. multiple development servers, each implementing a distinct Biolink Model version?). @newgene's proposal to nest tags under an `x-translator` block in the servers block, sounds sensible too.
